test/runtime/fs_full.c

```diff
@@ -0,0 +1,107 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/vfs.h>
+#include <assert.h>
+#include <errno.h>
+
+#define BIGDATA "/bigfile"
+#define NEWFILE "/newfile"
+
+int write_blocks(int fd, int nb)
+{
+    uint8_t buf[512];
+    int b;
+    for (b = 0; b < nb; b++) {
+        uint64_t *p = (uint64_t *)buf;
+        while (p < (uint64_t *)(buf + sizeof(buf)))
+            *p++ = b;
+        if (write(fd, buf, sizeof(buf)) <= 0) {
+            printf("failed writing at block %d err '%s'\n", b, strerror(errno));
+            assert(errno == ENOSPC);
+            break;
+        }
+        if (pread(fd, buf, sizeof(buf), 512 * b) != sizeof(buf)) {
+            printf("failed to read written block %d err '%s'\n", b, strerror(errno));
+            break;
+        }
+        p = (uint64_t *)buf;
+        while (p < (uint64_t *)(buf + sizeof(buf))) {
+            if (*p++ != b) {
+                printf("block %d does not match expected pattern in validation\n", b);
+                break;
+            }
+        }
+    }
+    return b;
+}
+
+int check_blocks(int fd, int nb)
+{
+    uint8_t buf[512];
+    int b;
+    for (b = 0; b < nb; b++) {
+        if (read(fd, buf, sizeof(buf)) <= 0) {
+            printf("failed reading at block %d err '%s'\n", b, strerror(errno));
+            break;
+        }
+        uint64_t *p = (uint64_t *)buf;
+        while (p < (uint64_t *)(buf + sizeof(buf))) {
+            if (*p++ != b) {
+                printf("block %d does not match expected pattern\n", b);
+                return b;
+            }
+        }
+    }
+    return b;
+}
+
+int main(int argc, char **argv)
+{
+    struct statfs statbuf;
+    int fd;
+
+    setbuf(stdout, NULL);
+    assert(statfs(argv[0], &statbuf) == 0);
+    uint64_t bfree = statbuf.f_bfree;
+    uint64_t btotal = statbuf.f_blocks;
+    printf("total bytes: %lu free bytes: %lu (actual: %lu)\n", btotal * 512, bfree * 512, statbuf.f_bfree*512);
```

Comment:
Isn't bfree the same as statbuf.f_bfree here?

Reply:
Yes--this is related leftover debugging I forgot to remove, so I'll take it out.